In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm
import re
import warnings
warnings.filterwarnings(action='ignore')

In [ ]:
! pip install impyute

In [ ]:
from impyute.imputation.cs import mice
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [ ]:
import lightgbm as lgb
from sklearn.metrics import *
from sklearn.model_selection import KFold

In [ ]:
train_err = pd.read_csv('/content/drive/MyDrive/데이콘/data/train_err_data.csv')
train_problem = pd.read_csv('/content/drive/MyDrive/데이콘/data/train_problem_data.csv')
test_err = pd.read_csv('/content/drive/MyDrive/데이콘/data/test_err_data.csv')
sample_submission = pd.read_csv('/content/drive/MyDrive/데이콘/data/sample_submission.csv')

In [ ]:
train_quality = pd.read_csv('/content/drive/MyDrive/데이콘/data/train_quality_data.csv')
test_quality = pd.read_csv('/content/drive/MyDrive/데이콘/data/test_quality_data.csv')

In [ ]:
train_err.head()

user_id            time model_nm       fwver  errtype errcode
0    10000  20201101025616  model_3  05.15.2138       15       1
1    10000  20201101030309  model_3  05.15.2138       12       1
2    10000  20201101030309  model_3  05.15.2138       11       1
3    10000  20201101050514  model_3  05.15.2138       16       1
4    10000  20201101050515  model_3  05.15.2138        4       0

In [ ]:
train_quality.head()

time  user_id       fwver  ...  quality_10  quality_11  quality_12
0  20201129090000    10000  05.15.2138  ...           4           0           0
1  20201129090000    10000  05.15.2138  ...           4           0           0
2  20201129090000    10000  05.15.2138  ...           4           0           0
3  20201129090000    10000  05.15.2138  ...           4           0           0
4  20201129090000    10000  05.15.2138  ...           4           0           0

[5 rows x 16 columns]

In [ ]:
train_quality['user_id'].value_counts()

15357    4224
23065    3828
22632    3744
11635    3504
16147    3336
         ... 
11014      12
24847      12
12282      12
18186      12
18698      12
Name: user_id, Length: 8281, dtype: int64

In [ ]:
train_quality[train_quality['user_id']==18698]

time  user_id       fwver  ...  quality_10  quality_11  quality_12
470784  20201106223000    18698  04.16.3553  ...           2           0           0
470785  20201106223000    18698  04.16.3553  ...           2           0           0
470786  20201106223000    18698  04.16.3553  ...           2           0           0
470787  20201106223000    18698  04.16.3553  ...           2           0           0
470788  20201106223000    18698  04.16.3553  ...           2           0           0
470789  20201106223000    18698  04.16.3553  ...           2           0           0
470790  20201106223000    18698  04.16.3553  ...           2           0           0
470791  20201106223000    18698  04.16.3553  ...           2           0           0
470792  20201106223000    18698  04.16.3553  ...           2           0           0
470793  20201106223000    18698  04.16.3553  ...           2           0           0
470794  20201106223000    18698  04.16.3553  ...           2           0           0
470795  20201106223000    18698  04.16.3553  ...           2           0           0

[12 rows x 16 columns]

In [ ]:
train_quality[train_quality['user_id']==11014]

time  user_id       fwver  ...  quality_10  quality_11  quality_12
52884  20201114014000    11014  04.33.1261  ...           4           0           0
52885  20201114014000    11014  04.33.1261  ...           4           0           0
52886  20201114014000    11014  04.33.1261  ...           4           0           0
52887  20201114014000    11014  04.33.1261  ...           4           0           0
52888  20201114014000    11014  04.33.1261  ...           4          -1           0
52889  20201114014000    11014  04.33.1261  ...           4          -1           0
52890  20201114014000    11014  04.33.1261  ...           4          -1           0
52891  20201114014000    11014  04.33.1261  ...           4           0           0
52892  20201114014000    11014  04.33.1261  ...           4           0           0
52893  20201114014000    11014  04.33.1261  ...           4           0           0
52894  20201114014000    11014  04.33.1261  ...           4           0           0
52895  20201114014000    11014  04.33.1261  ...           4           0           0

[12 rows x 16 columns]

* train_quality에서 user_id = 11014인 행만 추출한 결과 quality_0,1,2,5,6,11은 값이 같았고 quality10에서는 모두 4로 값이 같음을 알 수 있다. 

In [ ]:
train_quality[train_quality['user_id']==12282]

time  user_id       fwver  ...  quality_10  quality_11  quality_12
120552  20201117232000    12282  04.22.1750  ...           4          -1           0
120553  20201117232000    12282  04.22.1750  ...           4          -1           0
120554  20201117232000    12282  04.22.1750  ...           4          -1           0
120555  20201117232000    12282  04.22.1750  ...           4          -1           0
120556  20201117232000    12282  04.22.1750  ...           4          -1           0
120557  20201117232000    12282  04.22.1750  ...           4          -1           0
120558  20201117232000    12282  04.22.1750  ...           4          -1           0
120559  20201117232000    12282  04.22.1750  ...           4          -1           0
120560  20201117232000    12282  04.22.1750  ...           4           0           0
120561  20201117232000    12282  04.22.1750  ...           4           0           0
120562  20201117232000    12282  04.22.1750  ...           4           0           0
120563  20201117232000    12282  04.22.1750  ...           4           0           0

[12 rows x 16 columns]

In [ ]:
train_quality[train_quality['user_id']==20000]

time  user_id       fwver  ...  quality_10  quality_11  quality_12
535476  20201121021000    20000  05.15.2138  ...           1           0           0
535477  20201121021000    20000  05.15.2138  ...           1           0           0
535478  20201121021000    20000  05.15.2138  ...           1           0           0
535479  20201121021000    20000  05.15.2138  ...           1           0           0
535480  20201121021000    20000  05.15.2138  ...           1           0           0
535481  20201121021000    20000  05.15.2138  ...           1           0           0
535482  20201121021000    20000  05.15.2138  ...           1           0           0
535483  20201121021000    20000  05.15.2138  ...           1           0           0
535484  20201121021000    20000  05.15.2138  ...           1           0           0
535485  20201121021000    20000  05.15.2138  ...           1           0           0
535486  20201121021000    20000  05.15.2138  ...           1           0           0
535487  20201121021000    20000  05.15.2138  ...           1           0           0
535488  20201125021000    20000  05.15.2138  ...          55           0           0
535489  20201125021000    20000  05.15.2138  ...          55           0           0
535490  20201125021000    20000  05.15.2138  ...          55           0           0
535491  20201125021000    20000  05.15.2138  ...          55           0           0
535492  20201125021000    20000  05.15.2138  ...          55           0           0
535493  20201125021000    20000  05.15.2138  ...          55           0           0
535494  20201125021000    20000  05.15.2138  ...          55           0           0
535495  20201125021000    20000  05.15.2138  ...          55           0           0
535496  20201125021000    20000  05.15.2138  ...          55           0           0
535497  20201125021000    20000  05.15.2138  ...          55           0           0
535498  20201125021000    20000  05.15.2138  ...          55           0           0
535499  20201125021000    20000  05.15.2138  ...          55           0           0

[24 rows x 16 columns]

In [ ]:
train_quality['time'].dtype

dtype('int64')

In [ ]:
train_err_id_18696 = train_err[train_err['user_id']==18698]
train_err_id_18696[train_err_id_18696['time'] == 20201106223000] 

Empty DataFrame
Columns: [user_id, time, model_nm, fwver, errtype, errcode]
Index: []

* train_err와 train_quality 데이터의 id는 같아도 time이 겹치지는 않는다. 

In [ ]:
train_quality_id_15357 = train_quality[train_quality['user_id']==15357]
train_quality_id_15357['time'].value_counts() # 352개의 다른 time이 존재, 한 time당 12개씩 존재

train_err_id_15357 = train_err[train_err['user_id']==15357]
train_err_id_15357['time'].value_counts() # 한 time당 규칙적으로 개수가 일정하지 않음

id_15357_quality_list = train_quality_id_15357['time'].unique()
id_15357_err_list = train_err_id_15357['time'].unique()

intersection = list(set(id_15357_quality_list)&set(id_15357_err_list))
intersection # id=15357인 quality 데이터와 err데이터의 겹치는 time이 존재하지 않음

[]

* train_err_data와 train_quality_data의 user_id에 따른 time이 겹치지않는 데이터들이 존재합니다. 

In [ ]:
train_quality_id_16147 = train_quality[train_quality['user_id']==16147]
train_quality_id_16147['time'].value_counts() # 278개의 다른 time이 존재, 한 time당 12개씩 존재

train_err_id_16147 = train_err[train_err['user_id']==16147]
train_err_id_16147['time'].value_counts() # 한 time당 규칙적으로 개수가 일정하지 않음

id_16147_quality_list = train_quality_id_16147['time'].unique()
id_16147_err_list = train_err_id_16147['time'].unique()

intersection = list(set(id_16147_quality_list)&set(id_16147_err_list))
intersection # id=15357인 quality 데이터와 err데이터의 겹치는 time이 1개 존재함

[20201106081000]

* train_err_data와 train_quality_data의 user_id에 따른 time이 겹치는 데이터가 존재하기도 한다. 

In [ ]:
# train_quality의 time을 날짜 시간 초별로 구분하기
train_quality.rename(columns={"time": "original_time"}, inplace=True)
train_quality['DT'] = pd.to_datetime(train_quality['original_time'], format='%Y%m%d%H%M%S')
train_quality['date'] = train_quality['DT'].apply(lambda x: x.strftime('%Y-%m-%d'))
train_quality['time'] = train_quality['DT'].apply(lambda x: x.strftime('%H:%M:%S'))
train_quality['hour'] = train_quality['DT'].apply(lambda x: x.strftime('%H'))
train_quality.head()

original_time  user_id       fwver  ...        date      time  hour
0  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09
1  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09
2  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09
3  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09
4  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09

[5 rows x 20 columns]

In [ ]:
train_err.head()

user_id            time model_nm       fwver  errtype errcode
0    10000  20201101025616  model_3  05.15.2138       15       1
1    10000  20201101030309  model_3  05.15.2138       12       1
2    10000  20201101030309  model_3  05.15.2138       11       1
3    10000  20201101050514  model_3  05.15.2138       16       1
4    10000  20201101050515  model_3  05.15.2138        4       0

In [ ]:
train_err.rename(columns={"time": "original_time"}, inplace=True)
train_err['DT'] = pd.to_datetime(train_err['original_time'], format='%Y%m%d%H%M%S')
train_err['date'] = train_err['DT'].apply(lambda x: x.strftime('%Y-%m-%d'))
train_err['time'] = train_err['DT'].apply(lambda x: x.strftime('%H:%M:%S'))
train_err['hour'] = train_err['DT'].apply(lambda x: x.strftime('%H'))
train_err.head()

user_id   original_time model_nm  ...        date      time hour
0    10000  20201101025616  model_3  ...  2020-11-01  02:56:16   02
1    10000  20201101030309  model_3  ...  2020-11-01  03:03:09   03
2    10000  20201101030309  model_3  ...  2020-11-01  03:03:09   03
3    10000  20201101050514  model_3  ...  2020-11-01  05:05:14   05
4    10000  20201101050515  model_3  ...  2020-11-01  05:05:15   05

[5 rows x 10 columns]

In [ ]:
train_err[train_err['user_id'] == 10000]['original_time'].value_counts()

20201119223331    4
20201107110543    4
20201118163318    4
20201103140502    4
20201107070542    3
                 ..
20201110021658    1
20201122090120    1
20201120075804    1
20201113051136    1
20201115000313    1
Name: original_time, Length: 285, dtype: int64

In [ ]:
train_err[(train_err['original_time'] == 20201119223331) & (train_err['user_id'] == 10000)]

user_id   original_time model_nm  ...        date      time hour
213    10000  20201119223331  model_3  ...  2020-11-19  22:33:31   22
214    10000  20201119223331  model_3  ...  2020-11-19  22:33:31   22
215    10000  20201119223331  model_3  ...  2020-11-19  22:33:31   22
216    10000  20201119223331  model_3  ...  2020-11-19  22:33:31   22

[4 rows x 10 columns]

* train_err 데이터에서 user_id가 10000인 original_time이 같은 경우의 데이터프레임을 추출했다. 
    * 날짜시간이 모두 같은 데이터인데 errtype만 14,10,12,11으로 다르게 나왔다.
    * 같은 기기에서 동시간에 다른 errtype이 나올 수 있는 이유는 무엇일까?   


In [ ]:
train_err['model_nm'].value_counts()

model_1    5384491
model_0    4176279
model_2    3473254
model_3    1907438
model_4    1507802
model_8      36064
model_7      35356
model_5      32233
model_6       1746
Name: model_nm, dtype: int64

- 에러가 가장 많이 발생한 모델은 model_1이며 가장 적게 발생한 모델은 model_6이다. 

In [ ]:
train_err['errtype'].value_counts()

23    2276515
31    1985294
22    1865361
15    1604951
4     1296750
16    1249376
32    1130564
5      952920
40     730378
28     544306
27     540804
26     536401
12     320181
11     307030
14     252015
33     178477
34     165536
10     133403
41     104138
7       51993
6       50482
17      46009
24      39071
42      34374
3       30758
2       26106
13      22843
1       21545
39      15820
35      10187
37       9336
36       9317
38       4539
18       2671
20       2080
25       1184
19        664
21        607
30        402
8         166
9         109
Name: errtype, dtype: int64

- 에러가 가장 많이 발생한 errtypq은 23이고 가장 적게 발생한 errtype은 9이다.   

In [ ]:
train_err['errcode'].value_counts()

1                     8906967
0                     2599123
connection timeout    1835600
B-A8002                897863
80                     334018
                       ...   
14154                       1
6005                        1
10827                       1
3429                        1
8414                        1
Name: errcode, Length: 2805, dtype: int64

- errcode 값은 매우 다양한 형태로 나타났다. 
    * 차례로 숫자 1, 0, connection timeout이 많이 발생한 것을 알 수 있다.  

In [ ]:
train_err[train_err['errtype']==23]['date'].value_counts()

2020-11-07    100059
2020-11-01     97298
2020-11-08     96143
2020-11-15     90408
2020-11-06     82230
2020-11-28     82159
2020-11-02     81898
2020-11-09     81875
2020-11-14     79388
2020-11-05     79077
2020-11-03     78948
2020-11-10     78200
2020-11-04     77836
2020-11-22     74894
2020-11-29     73336
2020-11-12     72884
2020-11-11     72762
2020-11-16     72481
2020-11-13     70951
2020-11-21     70811
2020-11-23     70761
2020-11-17     69494
2020-11-19     69141
2020-11-18     68014
2020-11-27     67774
2020-11-20     67593
2020-11-26     67402
2020-11-25     63230
2020-11-30     60358
2020-11-24     59108
2020-12-01         2
Name: date, dtype: int64

In [ ]:
train_err[train_err['errtype']==23]['hour'].value_counts()

20    149707
21    147524
19    142923
22    137470
18    123348
23    118562
17    106342
16     96127
13     93742
12     92713
15     92011
00     90924
14     90505
09     89938
10     89280
11     88646
08     86004
01     72663
02     70801
03     67488
07     65812
04     61076
05     51706
06     51203
Name: hour, dtype: int64

In [ ]:
train_err[train_err['errtype']==31]['date'].value_counts()

2020-11-22    69939
2020-11-18    69795
2020-11-29    69270
2020-11-25    69200
2020-11-15    69042
2020-11-19    68840
2020-11-30    68093
2020-11-07    67370
2020-11-28    67065
2020-11-20    67040
2020-11-24    66873
2020-11-08    66860
2020-11-23    66638
2020-11-27    66614
2020-11-21    66098
2020-11-12    65925
2020-11-13    65687
2020-11-11    65340
2020-11-14    65170
2020-11-26    65153
2020-11-16    65126
2020-11-17    65088
2020-11-09    64869
2020-11-05    64689
2020-11-10    64636
2020-11-06    64021
2020-11-01    63810
2020-11-02    63010
2020-11-04    62287
2020-11-03    61745
2020-12-01        1
Name: date, dtype: int64

In [ ]:
train_err[train_err['errtype']==31]['hour'].value_counts()

18    130920
17    123603
19    122564
08    111260
20    107548
16    104983
07    104694
21    104151
09     96318
22     89679
15     89597
12     85926
13     84488
14     83419
10     82871
11     80087
23     65447
06     62428
00     47978
05     45654
04     43906
02     42861
03     41694
01     33218
Name: hour, dtype: int64

In [ ]:
train_err['hour'].value_counts()

19    899350
18    848590
20    848529
21    846325
22    774901
17    767757
00    722370
08    718565
16    701977
02    691168
03    686960
09    685861
23    684248
04    680424
13    660555
15    653461
07    653261
12    646229
10    643472
14    635799
11    629336
05    554346
06    470877
01    450302
Name: hour, dtype: int64

- 에러가 가장 많이 발생하는 시간대는 저녁(18~22)라는 것을 알 수 있고 가장 적게 발생하는 시간대는 새벽이나 오전인 것으로 나타났다. 

In [ ]:
train_err['date'].value_counts()

2020-11-21    649909
2020-11-22    596759
2020-11-28    594271
2020-11-15    591868
2020-11-24    590701
2020-11-01    589297
2020-11-29    588534
2020-11-07    587836
2020-11-08    581062
2020-11-14    574420
2020-11-06    558375
2020-11-19    557735
2020-11-23    550252
2020-11-18    539401
2020-11-09    539332
2020-11-02    539259
2020-11-20    533996
2020-11-25    531097
2020-11-05    529307
2020-11-27    528339
2020-11-03    528263
2020-11-26    526613
2020-11-12    526538
2020-11-10    523930
2020-11-16    522448
2020-11-04    522174
2020-11-11    519095
2020-11-17    516087
2020-11-30    512635
2020-11-13    505044
2020-12-01        78
2020-10-31         5
2020-12-02         3
Name: date, dtype: int64

In [ ]:
train_quality.head(2)

original_time  user_id       fwver  ...        date      time  hour
0  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09
1  20201129090000    10000  05.15.2138  ...  2020-11-29  09:00:00    09

[2 rows x 20 columns]

In [ ]:
train_quality['hour'].value_counts()

22    59352
20    54132
21    52440
23    46860
18    43104
19    42348
00    41808
14    41292
16    36408
12    36012
15    36012
10    35592
17    35052
13    32400
01    31080
11    30744
09    28260
02    27036
08    25704
03    22968
04    21024
07    17472
05    16440
06    15084
Name: hour, dtype: int64

- train_err 데이터셋과 비슷하게 quality 측정 시간은 저녁 (21~23시)때가 많고 새벽이 가장 적음을 알 수 있다. 

In [ ]:
# train_quality, train_err의 user_id가 겹치는가?
quality_user_id = train_quality['user_id'].unique()
err_user_id = train_err['user_id'].unique()
intersection = list(set(quality_user_id) & set(err_user_id))
len(intersection)

8281

- 에러가 발생한 user_id와 퀄리티를 측정한 user_id가 약 15000개 중 8281개가 겹친다. 

In [ ]:
len(quality_user_id)

8281

In [ ]:
train_quality_user_id = train_quality['user_id'].unique()
train_problem_user_id = train_problem['user_id'].unique()
intersection = list(set(train_quality_user_id)&set(train_problem_user_id))
len(intersection)

3167

In [ ]:
len(train_problem_user_id)

5000

In [ ]:
print(round(3176/5000*100),'%')

64 %


* 5000명의 불만을 제기한 유저 중 3167명 즉 약 64%가 quality 데이터셋에 존재한 유저이다.
  - 불만을 제기한 유저 중 절반이상이 quality 유저와 같다는 건 무엇을 의미할까?

In [ ]:
train_quality_df.isnull().sum()

user_id            0
quality_0     144432
quality_1          0
quality_2      40113
quality_3          0
quality_4          0
quality_5         20
quality_6          0
quality_7          0
quality_8          0
quality_9          0
quality_10         0
quality_11         0
quality_12         0
dtype: int64

* train_quality 데이터의 quality_0, quality_2, quality_5에 결측치가 약 14만,4만,20개가 존재한다.

In [ ]:
len(train_quality)

828624

In [ ]:
print(len(train_quality[train_quality['quality_0']==0]))
print(round(542790/828624*100),'%')
print(len(train_quality[train_quality['quality_0']==-1]))
print(round(130828/828624*100),'%')

542790
66 %
130828
16 %


* quality_0이 0인 갯수는 약 54만건이므로 데이터의 66%를 차지하고 -1인 갯수는 약 13만건으로 데이터의 16%를 차지한다. 
  * 즉, quality_0값이 0과 -1인 갯수의 합은 전체 데이터의 82%를 차지한다.


In [ ]:
print(len(train_quality[train_quality['quality_2']==0]))
print(round(632469/828624*100),'%')
print(len(train_quality[train_quality['quality_2']==-1]))
print(round(144392/828624*100),'%')

632469
76 %
144392
17 %


* quality_0과 마찬가지로 quality_2가 0과 -1인 갯수의 합은 전체 데이터의 무러 93%를 차지한다.

In [ ]:
print(len(train_quality[train_quality['quality_5']==0]))
print(round(428096/828624*100),'%')
print(len(train_quality[train_quality['quality_5']==-1]))
print(round(153354/828624*100),'%')

0
52 %
0
19 %


* quality_0과 quality_2와 마찬가지로 quality_5도 0과 -1인 갯수의 합은 전체 데이터의 무러 61%%를 차지한다. 

In [ ]:
print(len(test_quality[test_quality['quality_1']==0]))
print(round(576597/747972*100),'%')
print(len(test_quality[test_quality['quality_1']==-1]))
print(round(134061/747972*100),'%')

576597
77 %
134061
18 %


In [ ]:
print(len(test_quality[test_quality['quality_2']==0]))
print(round(578838/747972*100),'%')
print(len(test_quality[test_quality['quality_2']==-1]))
print(round(136864/747972*100),'%')

578838
77 %
136864
18 %


In [ ]:
print(len(test_quality[test_quality['quality_5']==0]))
print(round(399032/747972*100),'%')
print(len(test_quality[test_quality['quality_5']==-1]))
print(round(141788/747972*100),'%')

0
53 %
0
19 %


In [ ]:
test_quality_df.dtypes

user_id         int64
quality_0     float64
quality_1      object
quality_2     float64
quality_3       int64
quality_4       int64
quality_5      object
quality_6       int64
quality_7      object
quality_8      object
quality_9      object
quality_10     object
quality_11      int64
quality_12      int64
dtype: object